File: src/java/leetcode/editor/en/P589_NAryTreePreorderTraversal.java
package java.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import java.leetcode.editor.en.definition.*;

public class P589_NAryTreePreorderTraversal {
//leetcode submit region begin(Prohibit modification and deletion)
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> ans = new ArrayList<>();
        if (root == null) return ans;
        Stack<Node> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            ans.add(node.val);
            Collections.reverse(node.children);
            for (Node child : node.children)
                stack.push(child);
        }
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}

File: src/kt/leetcode/editor/en/P589_NAryTreePreorderTraversal.kt
package kt.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import kt.leetcode.editor.en.definition.*;

public class P589_NAryTreePreorderTraversal {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var children: List<Node?> = listOf()
 * }
 */

class Solution {
    fun preorder(root: Node?): List<Int> {
        val ans = ArrayList<Int>()
        /*  nested function begin */
        fun dfs(root: Node?) {
            if (root == null) return
            ans.add(root.`val`)
            root.children.forEach { dfs(it) }
        }
        /*  nested function end */
        dfs(root)
        return ans
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}
